ATMOS Pipe Repair and Replacement

Azrun opened the pull request:
[feature][input wanted][balance]

<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

1. Adding new state for simple atmos pipes. "Destroyed" 
   * All contents of the pipe segment (gas) will be passed to the turf/tile.
   * "Destroyed" state becomes the more likely state when **exploded** instead of pipe removal.
   * Destroyed state becomes a possible state when pipe pressure exceeds **10 times** the fatigue pressure.
![image](https://user-images.githubusercontent.com/1017374/107140995-d0318e80-68da-11eb-8136-69f6b1b4562c.png)

2. Destroyed state requires the piping segment be replaced by rods. 
   * 15 second duration (~7 seconds with carpenter or engineering training)
![rebuild_pipe](https://user-images.githubusercontent.com/1017374/107140857-fd317180-68d9-11eb-874f-fbdf6e5e9546.png)![rebuild_pipe_2](https://user-images.githubusercontent.com/1017374/107140867-028ebc00-68da-11eb-83f0-9a84fdbf8553.png)
   * 4 welds to fully repair
![rebuild_pipe_3](https://user-images.githubusercontent.com/1017374/107140870-0589ac80-68da-11eb-95f5-d31518bbe796.png)

3. Ruptured pipe **can** continue to rupture to a higher state. 

## Why's this needed? <!-- Describe why you think this should be added to the game. -->
Addresses issues and complaints about engineering getting bombed by providing engineering the capability of replacing severely damaged sections of pipe.

## Changelog
<!-- If necessary, put your changelog entry below. Otherwise, please delete it.
Use however you want to be credited in the changelog in place of CodeDude.
Use (*) for major changes and (+) for minor changes. For example: -->

```changelog
(u)Azrun
(*)Atmospheric pipes have a new severely damaged state that requires replacement by rods (and then welding).  Reduced chance of removing pipes from explosions in favor of new state.  Can also achieve new state through extreme pressure...
```

Azrun commented:
🤦 Need to play with pipe code to actually disconnect nodes due to how that logic is abstracted from reality, then mingling case won't be necessary.

pali6 commented:
Straigt up buildable pipes would be better (but also a large(r) project), though this is idea is seems fine as a temporary fix

Azrun commented:
> Straight up buildable pipes would be better (but also a large(r) project), though this is idea is seems fine as a temporary fix

Most of this change should be reverted/deleted it/when @ryebek or whoever actually gets around to constructable atmos. Exception being allowing rupture to escalate. 😈

predmetparfuyit86 commented:
> > Straight up buildable pipes would be better (but also a large(r) project), though this is idea is seems fine as a temporary fix
> 
> Most of this change should be reverted/deleted it/when @ryebek or whoever actually gets around to constructable atmos. Exception being allowing rupture to escalate. 😈

i think rybek has stopped working on that iirc

Azrun commented:
Added more "surface" to the destroyed pipes.  Improves ability to interact and place rods.  

Removed some "surface" from replaced pipes. Improves indication that additional action is required.

ZeWaka commented on code/modules/atmospherics/machinery/pipe.dm:
```diff
@@ -392,10 +443,14 @@ obj/machinery/atmospherics/pipe
-			if(node1&&node2)
+			if(destroyed)
+				icon_state = "destroyed"
+			else if(node1&&node2)
```
```suggestion
			else if(node1 && node2)
```

Azrun replied:
Propagated suggestion.

ZeWaka commented on code/modules/atmospherics/machinery/pipe.dm:
```diff
@@ -392,10 +443,14 @@ obj/machinery/atmospherics/pipe
 			..()
 
 		pipeline_expansion()
+			if(destroyed)
+				return list(null, null)
 			return list(node1, node2)
```
```suggestion
			. = list(node1, node2)
			if(destroyed)
				return list(null, null)
```

ZeWaka commented on code/modules/atmospherics/machinery/pipe.dm:
```diff
@@ -339,20 +344,35 @@ obj/machinery/atmospherics/pipe
-			else ruptured = 1
+		proc/rupture(pressure, destroy=FALSE)
+			var/new_rupture
+			if (src.destroyed || destroy || (pressure > 10*fatigue_pressure && prob(1)) )
```
ambig. grouping
```suggestion
			if (src.destroyed || destroy || ((pressure > (10 * fatigue_pressure)) && prob(1)))
```

ZeWaka commented on code/modules/atmospherics/machinery/pipe.dm:
```diff
@@ -339,20 +344,35 @@ obj/machinery/atmospherics/pipe
+				node2?.disconnect(src)
+				update_icon()
+				return
+			else if (pressure > 4*fatigue_pressure && prob(30)) new_rupture = 3
+			else if (pressure > 2*fatigue_pressure && prob(60)) new_rupture = 2
```
```suggestion
			else if ((pressure > (4 * fatigue_pressure)) && prob(30)) new_rupture = 3
			else if ((pressure > (2 * fatigue_pressure)) && prob(60)) new_rupture = 2
```

Azrun commented:
Testing functionality post addressing nits allowed me to find and address a bug! 🥳
